### DISCUSSION TITLE
Improve support for complex types in sqlite

### DISCUSSION THREAD

**Discussion by Sytten** [diesel-rs/diesel#3869] [Ideas]
Hi! I often need to create custom types with `ToSql` and `FromSql` to handle basic cases of writing say a `Vec` as `JSON` in sqlite databases. So say:
```rust
#[derive(Insertable, Clone)]
#[diesel(table_name = upstream_proxies)]
pub struct UpstreamProxyInsertSet {
    pub allowlist: Vec<String>, // Needs a wrapper
    pub denylist: Vec<String>,  // Needs a wrapper
}
```

So this got me thinking if diesel could help developers a bit further, I had (probably bad) ideas:
1. Feature flag to add `ToSql` and `FromSql` implementations for `Vec`, `HashMap`, etc that serialize to JSON. Maybe supporting a new `JSON` column type in the process to make sure that it matches.
2. Offer a `struct Json<T>(T)` wrapper structure to serialize to json. Though I am not a fan of wrapper objects for `Vec`, but it could implement all the Deref.
3. Custom `struct List<T>(Vec<T>)` structures. This is my least favorite option since it means losing the semantic and we need to re-implement a bunch of wrapping functions for the underlying object.
4. A `#[diesel(to_sql = func, from_sql = func)]` macro similar to what `serde` does when it (de)serializing unknown types. I like that a lot too since it can be applied to anything.

Thoughts?

**Comment by weiznich** [comment]
The fourth variant is already mostly implemented via the [`#[diesel(serialize_as)]`](https://docs.diesel.rs/2.1.x/diesel/prelude/derive.Insertable.html#optional-field-attributes) and [`#[diesel(deserialize_as)]`](https://docs.diesel.rs/2.1.x/diesel/prelude/derive.Queryable.html#optional-field-attributes) attributes. 

There is also [this unfinished PR](https://github.com/diesel-rs/diesel/pull/3837) that extends this functionality to also accept custom functions instead of only using the `From` trait. I think that's the right direction to go forward.

**Reply by Sytten** [reply]
That seems great! I didn't know about the `serialize_as`, that's exactly what I need.
